### PR TITLE
Introduce explicit API for configure test cluster feature flags

### LINKED ISSUE #83822: [CI] FullClusterRestartIT testSearchTimeSeriesMode failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/ljtyfyyt5g2rc/tests/:qa:full-cluster-restart:v8.0.0%23oldClusterTest/org.elasticsearch.upgrades.FullClusterRestartIT/testSearchTimeSeriesMode

**Reproduction line:**
`./gradlew ':qa:full-cluster-restart:v8.0.0#oldClusterTest' -Dtests.class="org.elasticsearch.upgrades.FullClusterRestartIT" -Dtests.method="testSearchTimeSeriesMode" -Dtests.seed=5D7C6139DC209842 -Dtests.bwc=true -Dtests.locale=no-NO -Dtests.timezone=Europe/Riga -Druntime.java=17`

**Applicable branches:**
8.0

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.upgrades.FullClusterRestartIT&tests.test=testSearchTimeSeriesMode

**Failure excerpt:**
```
org.elasticsearch.client.ResponseException: method [PUT], host [http://127.0.0.1:41499], URI [/testsearchtimeseriesmode], status line [HTTP/1.1 400 Bad Request]
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"unknown setting [index.routing_path] please check that any required plugins are installed, or check the breaking changes documentation for removed settings"}],"type":"illegal_argument_exception","reason":"unknown setting [index.routing_path] please check that any required plugins are installed, or check the breaking changes documentation for removed settings","suppressed":[{"type":"illegal_argument_exception","reason":"unknown setting [index.mode] did you mean [index.codec]?"}]},"status":400}

  at __randomizedtesting.SeedInfo.seed([5D7C6139DC209842:240C684B8CCEB1C7]:0)
  at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:346)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:312)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:287)
  at org.elasticsearch.upgrades.FullClusterRestartIT.createTimeSeriesModeIndex(FullClusterRestartIT.java:326)
  at org.elasticsearch.upgrades.FullClusterRestartIT.testSearchTimeSeriesMode(FullClusterRestartIT.java:231)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

### PR DESCRIPTION
**Pull request by mark-vieira**
Introduce improved handling for enabling feature flags on external test clusters. We are increasingly developing feature behind feature flags. For testing purposes, that means we need to set these flags when testing non-snapshot builds. This is the case both when a) doing release builds and b) when backward compatibility testing against released versions which require the feature flag for certain tests. Until now this was done via convoluted logic in the build scripts themselves. This pull request moves this logic into the test cluster orchestration itself. This means in our build scripts we can simply do this:

```groovy
requiresFeature 'es.index_mode_feature_flag_registered', Version.fromString("8.0.0")
```

This simply indicates that if we are attempting to start a non-snapshot release node with a version of 8.0.0 or later, we must pass the appropriate flag. 

Closes https://github.com/elastic/elasticsearch/issues/83822

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalTestClustersPlugin.java [hunk 1]
```diff
@@ -31,6 +32,10 @@ public void apply(Project project) {
         project.getRootProject().getPluginManager().apply(InternalReaperPlugin.class);
         TestClustersPlugin testClustersPlugin = project.getPlugins().apply(TestClustersPlugin.class);
         testClustersPlugin.setRuntimeJava(providerFactory.provider(() -> BuildParams.getRuntimeJavaHome()));
+        testClustersPlugin.setIsReleasedVersion(
+            version -> (version.equals(VersionProperties.getElasticsearchVersion()) && BuildParams.isSnapshotBuild() == false)
+                || BuildParams.getBwcVersions().unreleasedInfo(version) == null
+        );
```
Comment (nik9000) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalTestClustersPlugin.java:38]: :+1:

Comment (nik9000) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/InternalTestClustersPlugin.java:38]: "either it is the current version *and* we're running a release build OR it's not unreleased". That's a lot of double negatives, but I get it.

**Review by nik9000 — COMMENTED** [review]
I certainly like it.

File: x-pack/plugin/build.gradle [hunk 2]
```diff
@@ -152,11 +153,9 @@ testClusters.configureEach {
   extraConfigFile nodeCert.name, nodeCert
   extraConfigFile serviceTokens.name, serviceTokens
 
-  if (BuildParams.isSnapshotBuild() == false) {
-    systemProperty 'es.index_mode_feature_flag_registered', 'true'
-    systemProperty 'es.random_sampler_feature_flag_registered', 'true'
-    systemProperty 'es.user_profile_feature_flag_enabled', 'true'
-  }
+  requiresFeature 'es.index_mode_feature_flag_registered', Version.fromString("8.0.0")
```
Comment (pugnascotia) [inline x-pack/plugin/build.gradle:156]: Should this be 8.1.0 as well, like the ones below?

**Review by pugnascotia — COMMENTED** [review]
I wonder if we should abbreviate the usage i.e. not embed the full setting name?

```gradle
requiresFeature 'index_mode', Version.fromString("8.0.0")
```

**Comment by ywangd** [comment]
> Should we modify these to conform with the convention of the first set?

@mark-vieira For `es.user_profile_feature_flag_enabled`, Yes making it behave consistently with the first set sounds good to me. Thanks a lot for this initiative.

**Comment by tvernum** [comment]
The conversation on what our convention should be stalled last year, and is waiting for a new owner.

In the absence of a conclusion about that convention, my view was that the simplest option with `es.user_profile_feature_flag_enabled` was _not_ to fail if you tried to set it in a snapshot build.

Specifically, the primary reason for needing _this_ PR is because the model used in the other feature flags means that the build needs specific logic to know whether it is OK to set the flag. It's simpler to allow the setting all the time.

I don't mind changing `es.user_profile_feature_flag_enabled` to match the other flags, but we really need to make a decision about what our feature flag convention is supposed to look like (but we might need to wait until someone has capacity to own that).

**Comment by mark-vieira** [comment]
Ok, for now I've simply made `rollup_v2_feature_flag_enabled` follow the behavior of being auto-enabled on snapshot builds. This is _enough_ consistency for the testing stuff to work as intended. As @tvernum mentions we should have a longer discussion about how to implement feature flags. They should really be a first-class citizen in our settings/configuration model.

**Comment by pugnascotia** [comment]
Agreed, I'm confident that feature flags are going to become way more important in the future, so we want to get ahead of that in terms of support in the build and ES itself.

**Comment by mark-vieira** [comment]
@elasticmachine run elasticsearch-ci/full-bwc

**Pull request merged**